### PR TITLE
Allow token to expire

### PR DESCRIPTION
**Pull request by rullzer**
Requires:
- [x] https://github.com/nextcloud/server/pull/9484

Allows for tokens to be expired (for OAuth).

To make sure auth doesn't happen with expired tokens we thrown an ExpiredTokenException (that does hold the token). This seemed easier than adding a check everywhere we use it.

The tokens need to stay in the DB else we can't refresh them later.

### CODE REVIEW

**Review by ChristophWurst — APPROVED** [review]
Code looks good!

File: lib/private/Authentication/Token/DefaultTokenProvider.php [hunk 1]
```diff
@@ -165,10 +165,16 @@ public function getTokenByUser(IUser $user): array {
...
 	public function getToken(string $tokenId): IToken {
 		try {
-			return $this->mapper->getToken($this->hashToken($tokenId));
+			$token = $this->mapper->getToken($this->hashToken($tokenId));
 		} catch (DoesNotExistException $ex) {
 			throw new InvalidTokenException();
 		}
+
+		if ($token->getExpires() !== null && $token->getExpires() < $this->time->getTime()) {
+			throw new ExpiredTokenException($token);
```
Comment (MorrisJobke) [inline lib/private/Authentication/Token/DefaultTokenProvider.php]: Add this to the PHPDocs?

File: lib/private/Authentication/Token/DefaultTokenProvider.php [hunk 2]
```diff
@@ -180,10 +186,16 @@ public function getToken(string $tokenId): IToken {
...
 	public function getTokenById(int $tokenId): IToken {
 		try {
-			return $this->mapper->getTokenById($tokenId);
+			$token = $this->mapper->getTokenById($tokenId);
 		} catch (DoesNotExistException $ex) {
 			throw new InvalidTokenException();
 		}
+
+		if ($token->getExpires() !== null && $token->getExpires() < $this->time->getTime()) {
+			throw new ExpiredTokenException($token);
```
Comment (MorrisJobke) [inline lib/private/Authentication/Token/DefaultTokenProvider.php]: Same here.

**Review by MorrisJobke — APPROVED** [review]
Beside the little documentation nitpicks this looks fine 👍

File: lib/private/Authentication/Token/DefaultTokenProvider.php (hunk 2 above)
Comment (rullzer) [reply lib/private/Authentication/Token/DefaultTokenProvider.php]: Done..

**Pull request merged**